### ISSUE TITLE
Enable Java 8 library desugaring (Android Studio 4.0)

### ISSUE THREAD

**Issue by david-allison-1** [ankidroid/Anki-Android#7052]
Unsure if this will be accepted just yet as it requires a major Android Studio bump - probably fine

Desugaring allows for more Java 8 APIs to be usable.

> **Java 8 library desugaring in D8 and R8**
> Android Studio now includes support for using a number of Java 8 language APIs without requiring a minimum API level for your app.
> 
> Through a process called _desugaring_, the DEX compiler, D8, in Android Studio 3.0 and higher already provided substantial support for Java 8 language features (such as lambda expressions, default interface methods, try with resources, and more). In Android Studio 4.0, the desugaring engine has been extended to be able to desugar Java language APIs. This means that you can now include standard language APIs that were available only in recent Android releases (such as `java.util.streams`) in apps that support older versions of Android.
> 
> The following set of APIs is supported in this release:
> 
> * Sequential streams (`java.util.stream`)
> * A subset of `java.time`
> * `java.util.function`
> * Recent additions to `java.util.{Map,Collection,Comparator}`
> * Optionals (`java.util.Optional`, `java.util.OptionalInt` and `java.util.OptionalDouble`) and some other new classes useful with the above APIs
> * Some additions to `java.util.concurrent.atomic` (new methods on `AtomicInteger`, `AtomicLong` and `AtomicReference`)
> * `ConcurrentHashMap` (with bug fixes for Android 5.0)
> 
> To support these language APIs, D8 compiles a separate library DEX file that contains an implementation of the missing APIs and includes it in your app. The desugaring process rewrites your app’s code to instead use this library at runtime.
> 
> To enable support for these language APIs, include the following in your module’s `build.gradle` file:
> 
> ```groovy
> android {
>   defaultConfig {
>     // Required when setting minSdkVersion to 20 or lower
>     multiDexEnabled true
>   }
> 
>   compileOptions {
>     // Flag to enable support for the new language APIs
>     coreLibraryDesugaringEnabled true
>     // Sets Java compatibility to Java 8
>     sourceCompatibility JavaVersion.VERSION_1_8
>     targetCompatibility JavaVersion.VERSION_1_8
>   }
> }
> 
> dependencies {
>   coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.0.4'
> }
> ```

https://developer.android.com/studio/releases

**Comment by mikehardy** [comment]
I didn't think this required a major bump? We're already on 4.0.x - I typically follow within a week or so. I have no problem forcing to developers to move with new tools at all and usually just offer a week or so to do it. The tools themselves are really pester-y about updating anyway, why not take advantage of the features you get when you update? 4.1.0 is almost even - I've already handled a bug related to a gradle plugin change there, it's on rc02 so I expect we'll see it soon even.

**Comment by mikehardy** [comment]
This is potentially still useful, I just re-read it
Could allow us to move the main version of the google analytics dependency depending on audit of API support

https://developer.android.com/studio/write/java8-support#library-desugaring
https://developer.android.com/studio/write/java8-support-table

**Comment by Vaibhavi1707** [comment]
I apologize for the inconvenience caused as I'm working with gradle build files and CI for the first time. Thanks for the tip on comments will remember it in future. I have applied your changes and ran it on my end, where I ran into some issues.  Hi, I have faced an error in the Gradle build stating the maven is deprecated for gradle 7.0. It suggests to use the maven-publish plugin. On researching further I found that we'll have to replace the uploadArchives task with  generatePomFileForArchivesPublication to generate the POM XML
